src/interface/control_scriptbox.cpp

```diff
@@ -0,0 +1,496 @@
+
+          // Determine parameter rect
+          m_function_parameters.push_back({ Rectf(pos + SUGGESTION_DESCRIPTION_RECT_SPACING +
+                                                  Vector(name_width,
+                                                         i * Resources::control_font->get_height()),
+                                                  Sizef(Resources::control_font->get_text_width(param_name),
+                                                        Resources::control_font->get_height())),
+                                            std::move(param_description) });
+        }
+
+        if (i > 0)
+          out << std::setw(static_cast<int>(name_width / Resources::control_font->get_text_width(" "))) << ' ';
+        out << param_name;
+        if (i != static_cast<int>(func.parameters.size()) - 1)
+          out << ",\n";
+      }
+      out << ")";
+
+      m_suggestion_description.push_back(out.str());
+
+      desc_lines.push_back(func.description);
+      desc_lines.push_back(func.detailed_description);
+    }
+    break;
+
+    case ScriptingObject::Type::CLASS:
+    {
+      const auto& cl = static_cast<const ScriptingClass&>(*suggestion.reference);
+
+      if (suggestion.is_instance)
+        m_suggestion_description.push_back(fmt::format(fmt::runtime(_("Instance of `{}`.")), cl.name));
+
+      desc_lines.push_back(cl.summary);
+      if (!suggestion.is_instance)
+        desc_lines.push_back(cl.instances);
+    }
+    break;
+  }
+
+  for (std::string& p : desc_lines)
+  {
+    if (p.empty()) continue;
+
+    std::string result;
+    result += Resources::control_font->wrap_to_width(p, max_width, &p);
+    while (!p.empty())
+      result += '\n' + Resources::control_font->wrap_to_width(p, max_width, &p);
+
+    m_suggestion_description.push_back(std::move(result));
+  }
+
+  calculate_suggestion_description_rect();
+}
+
+void
+ControlScriptbox::autocomplete()
+{
+  assert(!m_suggestions.empty());
+
+  std::string line_content = get_contents(m_caret.line);
+  const size_t begin = line_content.find_first_not_of(" \t"); // Skip beginning spaces
+  line_content = line_content.substr(begin);
+
+  const std::string& suggestion = m_suggestions.at(m_selected_suggestion).name;
+
+  size_t line_pos = line_content.find_last_of('.');
+  if (line_pos == std::string::npos)
+    line_pos = 0;
+  else
+    line_pos++;
+
+  put_text(suggestion.substr(line_content.length() - line_pos));
+  parse_value();
+}
+
+void
+ControlScriptbox::on_caret_move()
+{
+  // To show suggestions, the caret must be at the end of a line
+  if (m_caret.line_pos < static_cast<int>(m_charlist[m_caret.line].size()))
+  {
+    m_suggestions.clear();
+    return;
+  }
+
+  std::string line_content = get_contents(m_caret.line);
+  const size_t begin = line_content.find_first_not_of(" \t"); // Skip beginning spaces
+  if (begin == std::string::npos)
+  {
+    m_suggestions.clear();
+    return;
+  }
+  line_content = line_content.substr(begin);
```

Comment:
This might be able to live in its own private, or even static function... just sayin'